File: python/asala/manifest.py
"""Manifest builder for creating content provenance."""
from datetime import datetime
from typing import Dict, Any, List, Optional

from .crypto import CryptoUtils
from .types import (
    ContentManifest,
    SignatureInfo,
    Assertion,
    ChainLink,
    ContentType,
)


class ManifestBuilder:
    """Builder for creating content manifests."""

    def __init__(
        self,
        content_hash: str,
        content_type: ContentType,
        creator: str
    ):
        """Initialize manifest builder.
        
        Args:
            content_hash: Hash of the content
            content_type: Type of content
            creator: Creator identifier
        """
        self._id = CryptoUtils.generate_uuid()
        self._content_hash = content_hash
        self._content_type = content_type
        self._created_at = datetime.utcnow().isoformat()
        self._created_by = creator
        self._signatures: List[SignatureInfo] = []
        self._assertions: List[Assertion] = []
        self._chain: List[ChainLink] = []

    def add_assertion(self, assertion_type: str, data: Dict[str, Any]) -> "ManifestBuilder":
        """Add an assertion to the manifest.
        
        Args:
            assertion_type: Type of assertion
            data: Assertion data
            
        Returns:
            Self for chaining
        """
        assertion = Assertion(
            type=assertion_type,
            data=data,
            hash=CryptoUtils.hash_string(CryptoUtils.canonical_json(data))
        )
        self._assertions.append(assertion)
        return self

    def add_metadata(self, metadata: Dict[str, Any]) -> "ManifestBuilder":
        """Add metadata assertion.
        
        Args:
            metadata: Metadata dictionary
            
        Returns:
            Self for chaining
        """
        return self.add_assertion("stds.metadata", metadata)

    def add_creation_info(
        self,
        device: str,
        software: str,
        location: Optional[Dict[str, float]] = None
    ) -> "ManifestBuilder":
        """Add creation information.
        
        Args:
            device: Device name
            software: Software used
            location: Optional location coordinates
            
        Returns:
            Self for chaining
        """
        return self.add_assertion("c2pa.created", {
            "device": device,
            "software": software,
            "location": location,
            "timestamp": datetime.utcnow().isoformat()
        })

    def sign(self, private_key: str, signer: str) -> "ManifestBuilder":
        """Sign the manifest.
        
        Args:
            private_key: Private key in PEM format
            signer: Signer identifier
            
        Returns:
            Self for chaining
        """
        manifest_data = {
            "id": self._id,
            "content_hash": self._content_hash,
            "created_at": self._created_at,
            "created_by": self._created_by,
            "assertions": [
                {"type": a.type, "data": a.data}
                for a in self._assertions
            ]
        }
        
        signature = CryptoUtils.sign_content(
            CryptoUtils.canonical_json(manifest_data),
            private_key
        )
        
        # Extract public key from private key for verification
        public_key = CryptoUtils.get_public_key_from_private_key(private_key)
        
        signature_info = SignatureInfo(
            algorithm="RSA-SHA256",
            public_key=public_key,
            signature=signature,
            timestamp=datetime.utcnow().isoformat(),
            signer=signer
        )
        
        self._signatures.append(signature_info)
        return self

    def add_chain_link(
        self,
        action: str,
        actor: str,
        private_key: str
    ) -> "ManifestBuilder":
        """Add a chain link for edits/transformations.
        
        Args:
            action: Action performed
            actor: Actor performing the action
            private_key: Private key for signing
            
        Returns:
            Self for chaining
        """
        previous_hash = (
            self._chain[-1].current_hash
            if self._chain
            else self._content_hash
        )
        
        link_data = {
            "action": action,
            "timestamp": datetime.utcnow().isoformat(),
            "actor": actor,
            "previous_hash": previous_hash
        }
        
        current_hash = CryptoUtils.hash_string(CryptoUtils.canonical_json(link_data))
        signature = CryptoUtils.sign_content(
            CryptoUtils.canonical_json(link_data),
            private_key
        )
        
        chain_link = ChainLink(
            action=action,
            timestamp=link_data["timestamp"],
            actor=actor,
            previous_hash=previous_hash,
            current_hash=current_hash,
            signature=signature
        )
        
        self._chain.append(chain_link)
        return self

    def build(self) -> ContentManifest:
        """Build the final manifest.
        
        Returns:
            Complete ContentManifest
        """
        return ContentManifest(
            id=self._id,
            content_hash=self._content_hash,
            content_type=self._content_type,
            created_at=self._created_at,
            created_by=self._created_by,
            signatures=self._signatures,
            assertions=self._assertions,
            chain=self._chain
        )

File: core/src/crypto/manifest.ts
import { CryptoUtils } from './utils';
import { ContentManifest, SignatureInfo, Assertion, ChainLink, ContentType } from '../types';

export class ManifestBuilder {
  private manifest: Partial<ContentManifest> = {};

  constructor(contentHash: string, contentType: ContentType, creator: string) {
    this.manifest = {
      id: this.generateId(),
      contentHash,
      contentType,
      createdAt: new Date().toISOString(),
      createdBy: creator,
      signatures: [],
      assertions: [],
      chain: []
    };
  }

  /**
   * Add an assertion to the manifest
   */
  addAssertion(type: string, data: Record<string, unknown>): this {
    const assertion: Assertion = {
      type,
      data,
      hash: CryptoUtils.hashString(JSON.stringify(data))
    };
    this.manifest.assertions!.push(assertion);
    return this;
  }

  /**
   * Add metadata assertion
   */
  addMetadata(metadata: Record<string, unknown>): this {
    return this.addAssertion('stds.metadata', metadata);
  }

  /**
   * Add creation info
   */
  addCreationInfo(device: string, software: string, location?: { lat: number; lng: number }): this {
    return this.addAssertion('c2pa.created', {
      device,
      software,
      location,
      timestamp: new Date().toISOString()
    });
  }

  /**
   * Sign the manifest
   */
  sign(privateKey: string, signer: string): this {
    const manifestData = {
      id: this.manifest.id!,
      contentHash: this.manifest.contentHash!,
      createdAt: this.manifest.createdAt!,
      createdBy: this.manifest.createdBy!,
      assertions: this.manifest.assertions
    };

    const signature = CryptoUtils.signContent(JSON.stringify(manifestData), privateKey);
    
    // Extract public key from private key for verification
    const publicKey = CryptoUtils.getPublicKeyFromPrivateKey(privateKey);
    
    const signatureInfo: SignatureInfo = {
      algorithm: 'RSA-SHA256',
      publicKey,
      signature,
      timestamp: new Date().toISOString(),
      signer
    };

    this.manifest.signatures!.push(signatureInfo);
    return this;
  }

  /**
   * Add chain link for edits/transformations
   */
  addChainLink(action: string, actor: string, privateKey: string): this {
    const previousHash = this.manifest.chain!.length > 0
      ? this.manifest.chain![this.manifest.chain!.length - 1].currentHash
      : this.manifest.contentHash!;

    const linkData = {
      action,
      timestamp: new Date().toISOString(),
      actor,
      previousHash
    };

    const currentHash = CryptoUtils.hashString(JSON.stringify(linkData));
    const signature = CryptoUtils.signContent(JSON.stringify(linkData), privateKey);

    const chainLink: ChainLink = {
      ...linkData,
      currentHash,
      signature
    };

    this.manifest.chain!.push(chainLink);
    return this;
  }

  /**
   * Build the final manifest
   */
  build(): ContentManifest {
    return this.manifest as ContentManifest;
  }

  private generateId(): string {
    return `urn:uuid:${CryptoUtils.hashString(Date.now().toString() + Math.random())}`;
  }
}
